core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/StrictEvaluationStrategy.java

```diff
@@ -654,26 +555,151 @@ protected BindingSet convert(Statement st) {
 						resultingIterator.close();
 					}
 				} finally {
-					try {
-						if (stIter3 != null) {
-							stIter3.close();
-						}
-					} finally {
-						try {
-							if (stIter2 != null) {
-								stIter2.close();
-							}
-						} finally {
-							if (stIter1 != null) {
-								stIter1.close();
-							}
-						}
+					if (stIter1 != null) {
+						stIter1.close();
 					}
 				}
 			}
 		}
 	}
 
+	/**
+	 * Generate a predicate that tests for Named contexts are matched by retrieving all statements from the store and
+	 * filtering out the statements that do not have a context. Or. The same variable might have been used multiple
+	 * times in this StatementPattern, verify value equality in those cases.
+	 */
+	protected Predicate<Statement> filterOnStatements(StatementPattern statementPattern, final Var subjVar,
+			final Var predVar,
+			final Var objVar, final Var conVar, final Value subjValue, final Value predValue, final Value objValue,
+			Resource[] contexts) {
+		Predicate<Statement> filter = null;
+		if (contexts.length == 0 && statementPattern.getScope() == Scope.NAMED_CONTEXTS) {
+			filter = (st) -> st.getContext() != null;
+		}
+		filter = filterSameVariable(subjVar, predVar, objVar, conVar, subjValue, predValue, objValue, filter);
+		return filter;
```

Comment:
I think I have a slight preference for moving this "chaining" of predicates back into the main method - for readability mostly. Also makes it easier to explain what this particular method does, I think.

Reply:
I like to keep it out. Because further down the line I want to separate this logic into a preparation stage and a execution stage. This would be part of the preparation stage, while the main try catch of the main body would be the execution stage.